modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/components/A11y/A11yChecker.ts

```diff
@@ -0,0 +1,398 @@
+
+			return this.nextFramePhase >= currentTime;
+		}
+
+		// There's still time left in the frame.
+
+		return false;
+	}
+
+	private async workLoop() {
+		this.currentTask = this.queue.peek();
+
+		while (this.currentTask !== undefined) {
+			if (this.shouldYieldToHost()) {
+
+				// We no longer have time to run currentTask now
+
+				break;
+			}
+
+			await this.currentTask.callback(this.currentTask.target);
+
+			this.queue.dequeue();
+
+			this.currentTask = this.queue.peek();
+
+			this.nextFramePhase = performance.now() + this.yieldInterval;
+		}
+
+		// Return whether there's additional work
+
+		return this.currentTask !== undefined;
+	}
+
+	scheduleCallback(callback: Function, target: T) {
+		const newTask = {
+			callback,
+			id: this.taskIdCounter++,
+			target,
+		};
+
+		this.queue.enqueue(newTask);
+
+		if (!this.isCallbackScheduled && !this.isPerformingWork) {
+			this.isCallbackScheduled = true;
+			this.requestHostCallback();
+		}
+	}
+}
+
+type Mutation = {
+	attributes: Record<string, string>;
+	nodeName: string;
+};
+
+export interface A11yCheckerOptions {
+
+	/**
+	 * AxeOptions is the set of settings that are passed to axe-core.
+	 */
+	axeOptions?: RunOptions;
+
+	/**
+	 * Callback for when the analysis is finished.
+	 */
+	callback: (violations: AxeResults) => void;
+
+	/**
+	 * Denylist is an optional list with CSS selectors that will be excluded
+	 * from the analysis.
+	 */
+	denylist?: Array<string>;
+
+	/**
+	 * Mutation is an optional list of criteria on which a new analysis will be
+	 * triggered.
+	 */
+	mutations?: Record<MutationRecordType, Mutation>;
+
+	/**
+	 * Targets is a list or element that represents the subtree(s) to be
+	 * analyzed.
+	 */
+	targets: Array<string>;
+}
+
+export class A11yChecker {
+	private callback: A11yCheckerOptions['callback'];
+	private scheduler: Scheduler<Node>;
+	private observers: Array<{
+		target: string;
+		mutation: MutationObserver;
+	}>;
+	private mutations?: Record<MutationRecordType, Mutation>;
+	readonly axeOptions: RunOptions;
+	readonly denylist: A11yCheckerOptions['denylist'];
+
+	constructor({
+		axeOptions,
+		callback,
+		denylist,
+		mutations,
+		targets,
+	}: A11yCheckerOptions) {
+		const defaultOptions = {
+			absolutePaths: true,
+			reporter: 'v2',
+		} as const;
+
+		this.axeOptions = axeOptions ? axeOptions : defaultOptions;
+
+		this.callback = callback;
+		this.denylist = denylist;
+
+		// Scheduler can be a singleton to allow multiple instances of
+		// A11yChecker to share the same queue to avoid bottlenecks and
+		// performance issues.
+
+		this.scheduler = new Scheduler();
+
+		this.mutations = mutations;
+
+		this.observers = targets.map((target: string) => ({
+			mutation: new MutationObserver(this.mutationCallback.bind(this)),
+			target,
+		}));
+	}
+
+	private async run(target: HTMLElement) {
+
+		// Only perform the analysis if the element exists on the screen
+
+		if (!document.body.contains(target)) {
```

Comment:
I understood the comment as `isInViewport` which based on this, not what it means at all 😂 

If we're seeing perf bottlenecks, maybe batching elements outside of viewport separately could reduce workload a little as long as the `isInViewport` check is cheaper than the `axe` one 🤔

Reply:
Oh 😅, it makes sense too. haha i think the comment got confused, maybe "exists on the DOM" is clearer.

> Then, I thought that if something was being heavy, you could scan in batches, first elements inside the viewport, then elements outside of it and keep a stream of results coming. That way you would first see elements with issues in the visible viewport get highlighted and then work would continue for those outside the visible area. It's just a premature optimization, so probably no need to dwell on this 😂

Yeah, it makes sense, that would be almost a priority system haha, we can think about it later.